ENH: make clipping unnecessary for entropy

DerWeh opened the pull request:
This drops (or for now deprecates) the clipping in entropy.
It was discussed with @huiwengoh in https://github.com/cleanlab/cleanlab/pull/689#issuecomment-1530536858.
The entropy is clearly defined that for a probability $p_i=0$, the limit is to be taken $p_i \log(p_i)=0$. This is implemented by using `np.sum(..., where=finite)` excluding vanishing terms from the sum. Furthermore, we use `np.log(..., where=finite)` not to calculate the logarithm. (So we still leave the terms in the product `*`.)

What remains to be done is add a deprecation message to the docstring: https://numpydoc.readthedocs.io/en/latest/format.html#deprecation-warning

Do you have any guidelines for this?

Note that I used `finite = pred_probs != 0` instead of `finite = pred_probs > 0` on purpose. The second version, why hide negative probabilities, which we do not check for. Thus, negative probabilities still throw a `RuntimeWarning` for invalid values in log, and return NaN. This seems like a meaningful behavior to me.

**Update**: I just found out about [scipy.special.xlogy](https://docs.scipy.org/doc/scipy/reference/generated/scipy.special.xlogy.html) which is even more readable.
Do you care to add `scipy` as dependency? This is no real further dependency as `scikit-learn>=1.0.0` depends already on `scipy>=1.1.0`, so it will be installed anyway.
Then we can use the shorter:
```python
entropy = -np.sum(xlogy(pred_probs, pred_probs), axis=1)
```
instead of the current version:
```python
finite = pred_probs != 0
entropy = -np.sum(pred_probs * np.log(pred_probs, where=finite), axis=1, where=finite)
```
Internally, it does the same.

----

We could add more thorough hypothesis tests if we add the development dependency `hypothesis[numpy]`. We could check if we get entropy in [0.0, 1.0] for any non-negative array inputs.

DerWeh commented:
The reason for test failure seems to be an outdated version of NumPy.
(Note, that not the changed function is broken for old NumPy versions, but the test itself cannot be run.)
Have no clear idea, but https://github.com/numpy/numpy/issues/22113 hints that there were issues in older NumPy versions.

My suggestion would be simply skipping the values if they are not available. Not really elegant, but we get the full test on new Python versions, and old versions shouldn't fail anymore.

huiwengoh commented on tests/test_outlier.py:
```diff
@@ -109,6 +109,19 @@ def make_data(
 data = make_data()
 
 
+def test_normalized_entropy():
```
I believe this function will belong better in `test_util.py`. 

As for the failing tests, a nice way to ensure it does not throw an error would be to use `val = getattr(info, "attribute", None)` to check if each attribute exists, and skip the test if `val is None`?

DerWeh replied:
I moved the test as requested. `getattr` is indeed a good workaround. Instead of skipping `None`, I provide a reasonable fallback right-away, keeping the code simpler.

huiwengoh commented:
The scipy method looks good! And yes we have some other internal functions already using scipy, so feel free to swap to the scipy method that looks much cleaner.

Re deprecation, it seems like we have not added deprecation docstrings hence there are no existing guidelines, following that numpy guide to use the `Deprecated` Sphinx directive sounds good tho!

DerWeh commented:
I updated the code to use `xlogy`, and added the deprecation role using [scipy.linalg.solve](https://github.com/scipy/scipy/blob/v1.10.1/scipy/linalg/_basic.py#L40-L265) as a reference.
For the version number, I used the next minor release (2.5.0); feel free to edit this.

huiwengoh commented:
Thanks for the changes - they look good! A couple of quick notes:
- I removed the explicitly stated scipy dependency, we've decided to not list it as it adds extra overhead to ensure that we keep it up to date with the version required with sklearn going forward. We also have ongoing CI tests to ensure that we have all the dependencies required for cleanlab so we should be able to catch any errors if sklearn decides to drop the scipy dependency in the future
- I think using the next minor release (2.5.0) is fine for now, we can update that later before the next release if anything changes
- I had one more small ask before approving and merging the PR which is to have some sort of error catching for invalid `pred_probs` values, otherwise everything else looks good!

DerWeh commented on cleanlab/internal/label_quality_utils.py:
```diff
@@ -95,21 +95,32 @@ def get_normalized_entropy(
     # Note that dividing by log(num_classes) changes the base of the log which rescales entropy to 0-1 range
-    clipped_pred_probs = np.clip(pred_probs, a_min=min_allowed_prob, a_max=None)
-    return -np.sum(pred_probs * np.log(clipped_pred_probs), axis=1) / np.log(num_classes)
+    return -np.sum(xlogy(pred_probs, pred_probs), axis=1) / np.log(num_classes)
```
You are right, this might be breaking behavior. Before, negative probabilities were silently ignored, as they were clipped to a tiny positive value.

Instead of an `assert`, I would rather raise a `ValueError` which seems suitable in this case. Assertion are not thrown if optimizations are turned on, and typically don't belong in production code.
The question follow-up question is, whether we should also catch probabilities larger than zero. While the formula is mathematically still well-defined, its meaning is questionable.

Another option would be simply rising a `RuntimeWarning` as `numpy.log` does for negative numbers (actually, I assumed that `xlogy` would behave the same way, but I forgot to check it).

huiwengoh replied:
Yes a `ValueError` sounds good. I think adding a check to ensure that probabilities are not larger than 1 is a good idea as well, good catch there!

huiwengoh reviewed (approved):
LGTM! Thanks for the contribution!
